src/parser/expression.js

```diff
@@ -1163,26 +1215,33 @@ export default class ExpressionParser extends LValParser {
 
   parseIdentifier(liberal?: boolean): N.Identifier {
     const node = this.startNode();
+    const name = this.parseIdentifierName(node.start, liberal);
+    node.name = name;
+    node.loc.identifierName = name;
+    return this.finishNode(node, "Identifier");
+  }
+
+  parseIdentifierName(pos: number, liberal?: boolean): string {
     if (!liberal) {
       this.checkReservedWord(this.state.value, this.state.start, !!this.state.type.keyword, false);
     }
 
+    let name: string;
+
     if (this.match(tt.name)) {
-      node.name = this.state.value;
+      name = this.state.value;
     } else if (this.state.type.keyword) {
-      node.name = this.state.type.keyword;
+      name = this.state.type.keyword;
     } else {
-      this.unexpected();
+      throw this.unexpected();
```

Comment:
Flow will complain without the `throw` since it's not clear that control flow can't reach the end without assigning `name`.